test/modules/DirectMethodReceiver/src/DirectMethodReceiver.cs

```diff
@@ -0,0 +1,106 @@
+            this.moduleClient = moduleClient;
+            this.testResultReportingClient = testResultReportingClient;
+            this.batchId = Guid.NewGuid();
+        }
+
+        public static async Task<DirectMethodReceiver> CreateAsync(
+            ILogger logger,
+            IConfiguration configuration)
+        {
+            Preconditions.CheckNotNull(logger, nameof(logger));
+            Preconditions.CheckNotNull(configuration, nameof(configuration));
+
+            ModuleClient moduleClient = await ModuleUtil.CreateModuleClientAsync(
+                configuration.GetValue("ClientTransportType", TransportType.Amqp_Tcp_Only),
+                ModuleUtil.DefaultTimeoutErrorDetectionStrategy,
+                ModuleUtil.DefaultTransientRetryStrategy,
+                logger);
+
+            TestResultReportingClient testResultReportingClient = null;
+            Option<Uri> testReportCoordinatorUrl = Option.Maybe(configuration.GetValue<Uri>("testResultCoordinatorUrl"));
+            if (testReportCoordinatorUrl.HasValue)
+            {
+                testResultReportingClient = new TestResultReportingClient
+                {
+                    BaseUrl = testReportCoordinatorUrl.Expect(
+                        () => new ArgumentException("testReportCoordinatorUrl is empty"))
+                        .AbsoluteUri
+                };
+            }
+
+            return new DirectMethodReceiver(
+                logger,
+                configuration,
+                moduleClient,
+                testResultReportingClient);
+        }
+
+        public void Dispose() => this.moduleClient?.Dispose();
+
+        async Task<MethodResponse> HelloWorldMethodAsync(MethodRequest methodRequest, object userContext)
+        {
+            this.logger.LogInformation("Received direct method call.");
+            // Send the report to Test Result Coordinator
+            await this.ReportTestResult();
+            return new MethodResponse((int)HttpStatusCode.OK);
+        }
+
+        public async Task ReportTestResult()
+        {
+            DirectMethodTestResult testResult = null;
+            if (this.testResultReportingClient != null)
+            {
+                testResult = new DirectMethodTestResult(this.configuration.GetValue<string>("IOTEDGE_MODULEID") + ".receive", DateTime.UtcNow)
+                {
+                    TrackingId = Option.Maybe(this.configuration.GetValue<string>("trackingId"))
+                        .Expect(() => new ArgumentException("TrackingId is empty")),
+                    BatchId = this.batchId.ToString(),
+                    SequenceNumber = this.directMethodCount.ToString(),
+                    Result = HttpStatusCode.OK.ToString()
+                };
+            }
+
+            await ModuleUtil.ReportTestResultAsync(this.testResultReportingClient, this.logger, testResult);
+            this.directMethodCount++;
+        }
+
+        public async Task StartAsync()
```

Comment:
I think it is better to move it inside constructor.

Reply:
constructor cannot be async.